accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+
+## Live Debugging
+
+Live debugging support comprises several features, each of which comes with
+design and implementation considerations. These features are: setting
+breakpoints, single stepping, watch points, step into/out of, and integration
+with our existing baseline compiler Winch.
+
+### Breakpoints and Single-stepping
+
+To implement breakpoints and single-stepping, we need the ability to pause
+program execution and wait for user input prior to executing a specific line of
+code. The way this is handled in native debuggers is as follows:
+
+* When a user sets a breakpoint, the instruction corresponding to that line of
+  code is replaced with an `int3` instruction.
+
+* When executed, the `int3` instruction raises a `SIGTRAP`, halting execution
+  and yielding control back to the debugger.
+
+This approach leads to a problem for us: we only have one compiled copy of a
+module’s code, even though multiple instances may be running the code. So if we
+just replace an instruction in the code with an interrupt, it is possible that a
+different instance will hit that interrupt. At that point, in order for that
+instance to continue executing we need to swap in the actual instruction. This
+leads to a race condition between the instances where we aren’t guaranteed that
+our breakpoint will actually interrupt when we need it to! We can get around
+this in the following way:
+
+1. When code is compiled for debugging, for every instruction emitted, emit a
+call to a utility function that can check if execution should be paused:
+
+2. When a breakpoint is set, the function will pause execution, allowing for
+resumption later on
+
+3. When an instance hits that breakpoint it can automatically continue if it
+isn’t intended to stop, otherwise it can wait for user input.
+
+We will need to keep track of some information for every wasm opcode in our
+program: what program counter the opcode is, the size of its operand stack, and
+which entry in the operand stack the opcode is in terms of machine state. All of
+this information can be serialized and passed into the utility function present
+in the `<check for breakpoint>` pseudo opcodes above.
+
+### Watchpoints
+
+Watchpoints for locals, globals, and table elements can be converted to
+breakpoints by statically checking for local.set, global.set, and table.set
+instructions. Memory watchpoints— pausing execution when a program writes to a
+specific memory location— are a little bit trickier to accomplish. However, like
+the proposed implementation for single-stepping mentioned in the [previous
+section](#breakpoints-and-single-stepping), we can implement watch-points by
+instrumenting all store operations with a similar call out to a utility
+function. The call sites (store instructions) would have enough local
+information available to determine if a watchpoint would trigger, and could
+provide us with another implementation path that would avoid modifying the
+signal handler.
```

Comment:
Another approach is protecting the memory where the variable is being watched so it's read-only, and handling the fault in user mode, if possible (e.g. `userfaultfd(2)` on Linux).  Maybe for Future Work as a possible optimization?